man/man8/abbrev.8
.TH ABBREV 8
.\" @(#) $Header: /xtel/pp/pp-beta/man/man8/RCS/abbrev.8,v 6.0 1991/12/18 20:44:03 jpo Rel $
.\"
.\" $Log: abbrev.8,v $
.\" Revision 6.0  1991/12/18  20:44:03  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
abbrev \- Abbreviates host\/domain names
.SH SYNOPSIS
abbrev local-domain1 [local-domain2 ...] < input-file
.SH DESCRIPTION
.I Abbrev
is a program that is used to construct the
.I domain
table residing in the PP table directory (tbldfldir).
This is a useful tool as it allows a user an alternative
way of specifying an RFC822 address.
If
.I abbrev
is used, an abbreviated forms of domains may be specified
which will then get properly expanded by the PP mail system.
For example an address "Jim.Hacker@cs" could be expanded
to "Jim.Hacker@cs.ucl.ac.uk", depending on
.I abbrev's
parameters.
.PP
.I Abbrev
takes a number of command line arguments, which should be local domains.
It takes as input a typical domain table format.
.PP
For each entry specified in the input-file,
.I abbrev
recursively shortens the key until the domain specified on the command
line and the given entry share no common parts. Each entry is
abbreviated with respect to each command line option.
.PP
Blank lines and comments are passed though unchanged.
.SH "SEE ALSO"
\fIThe PP Manual: Volume 1 \- Installation and Operation\fP


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Tools/tables/abbrev/abbrev.c
/* abbrev.c: abbreviates domains */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Tools/tables/abbrev/RCS/abbrev.c,v 6.0 1991/12/18 20:33:28 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Tools/tables/abbrev/RCS/abbrev.c,v 6.0 1991/12/18 20:33:28 jpo Rel $
 *
 * $Log: abbrev.c,v $
 * Revision 6.0  1991/12/18  20:33:28  jpo
 * Release 6.0
 *
 */



#include	"util.h"

extern char	*strdup (), *index();
static int	LineCount;
static char	*myname;
static char	*aliases[100];
static int	maxalias = 0;
static int	bitmap;

static void do_alias ();
/* ---------------------  Begin	 Routines  -------------------------------- */


main (ac,av)
int	ac;
char	*av [];
{
	char	buffer [BUFSIZ], *entries[10];
	register char	*alias, *name, *nonwildcard;
	register int	i, NumEntries, use;
	char	rhs[BUFSIZ], entry[BUFSIZ];

	myname = av [0];
	if (ac < 2) {
		fprintf (stderr, "Usage: %s Domain...\n", myname);
		exit (1);
	}


	for (i = 1; i < ac; i++) {
		(void) sprintf (buffer, "X.%s", av[i]);
		aliases[maxalias++] = strdup (buffer);
	}
	LineCount = 0;

	while (fgets (buffer, sizeof buffer, stdin) != NULL) {
		LineCount++;
		alias = &buffer [0];
		printf ("%s", alias);
		if (*alias == '#' || *alias == '\n')
			continue;

		alias [strlen (alias) - 1] = '\0';
		for (name = alias; *name != '\0'; name++)
			if (*name == ':' || *name == '#')
				break;
		if (*name == '#')
			continue;

		if (*name == '\0') {
			fprintf (stderr, "%s: Syntax Error [%s] ",
				myname, alias);
			fprintf (stderr, "at line %d - Ignored\n", LineCount);
			continue;
		}
		*name++ = '\0';

		if (alias[0] == '*'
		    && alias[1] == '.')
			nonwildcard = &(alias[2]);
		else
			nonwildcard = &(alias[0]);

		if ((NumEntries = sstr2arg (name, 10, entries, "|")) < 1) {
			fprintf(stderr, "%s: Syntax Error [%s] ",
				myname, alias);
			fprintf(stderr, "at line %d - Ignored\n", LineCount);
			continue;
		}
		rhs[0] = '\0';
		use = 1;
		for (i = 0; i < NumEntries && use == 1; i++) {
			use = do_entry(entries[i], nonwildcard, entry, alias);
			if (rhs[0] != '\0')
				strcat (rhs, "|");
			strcat (rhs, entry);
		}
		if (use == 1)
			for (i = 0; i < maxalias; i++)
				do_alias (alias, rhs, aliases[i]);
	}
	exit (0);
}

do_entry(name, nonwildcard, rhs, alias)
char	*name;
char 	*nonwildcard;
char	*rhs;
char	*alias;
{
	char	*vec[10];
	int	vecp, i;
	int	use = 0;
	int	gotnorm;
	int	mtaislhs;
	if ((vecp = sstr2arg (name, 10, vec, " \t")) < 1) {
		fprintf(stderr, "%s: Syntax Error [%s] ",
			myname, alias);
		fprintf(stderr, "at line %d - Ignored\n", LineCount);
		return use;
	}
		
	rhs[0] = '\0';
	gotnorm = 0;
	mtaislhs = 0;
	for (i = 0; i < vecp; i++) {
		char *p;

		if (lexnequ (vec[i], "norm+mta", 8) == 0
		      || lexnequ (vec[i], "mta+norm", 8) == 0) { 
			use = 1;
			gotnorm = 1;
			if (rhs[0] != '\0')
				strcat(rhs, " ");
			if ((p = index(vec[i], '=')) != (char *) 0
			    && *(p+1) != NULL)
				strcat(rhs, vec[i]);
			else {
				/* construct norm+mta= */
				strcat(rhs, "norm+mta=");
				strcat(rhs, nonwildcard);
			}
		} else if (lexnequ (vec[i], "norm", 3) == 0) {
			use = 1;
			gotnorm = 1;
			if (rhs[0] != '\0')
				strcat(rhs, " ");
			if ((p = index(vec[i], '=')) != (char *) 0
			    && *(p+1) != NULL)
				strcat(rhs, vec[i]);
			else {
				/* construct norm= */
				strcat(rhs, "norm=");
				strcat(rhs, nonwildcard);
			}
		} else if (lexnequ (vec[i], "synonym", 7) == 0) {
			gotnorm = 1;
			use = 1;
			if (rhs[0] != '\0')
				strcat(rhs, " ");
			strcat(rhs, vec[i]);
		} else if (lexnequ (vec[i], "local", 5) == 0) {
			use = 1;
			if (rhs[0] != '\0')
				strcat(rhs, " ");
			strcat(rhs, vec[i]);
		} else if (lexnequ (vec[i], "mta", 3) == 0) {
			use = 1;
			if (rhs[0] != '\0')
				strcat(rhs, " ");
			if ((p = index(vec[i], '=')) != (char *) 0
			    && *(p+1) != NULL)
				strcat(rhs, vec[i]);
			else 
				/* construct norm= */
				mtaislhs = 1;
		} else {
			if (rhs[0] != '\0')
				strcat(rhs, " ");
			strcat(rhs, vec[i]);
		}
	}
	bitmap = 0;

	if (gotnorm == 0) {
		char	*key;
		/* add norm=lhs */
		if (mtaislhs == 1) 
			key = "norm+mta=";
		else
			key = "norm=";
		if (rhs[0] != '\0')
			strcat(rhs, " ");
		strcat(rhs, key);
		strcat(rhs, nonwildcard);
	} else if (mtaislhs == 1) {
		if (rhs[0] != '\0')
			strcat(rhs, " mta=");
		else
			strcat(rhs, "mta=");
		strcat(rhs, nonwildcard);
	}
	return use;
}

/* ---------------------  Static  Routines  ------------------------------- */



static void do_alias (alias, name, domain)
register char	*alias, *name, *domain;
{
	register char	*a, *d;
	int dn = -1;

	a = &alias [strlen (alias) - 1];
	d = &domain [strlen (domain) - 1];
	for (; a != alias && *a == *d; a--, d--)
		if (*a == '.') {
			dn ++;
			if (bitmap & (1 << dn))
				continue;
			bitmap |= (1 << dn);
			*a = '\00';
			if (strcmp (alias, "*") != 0)
				printf ("%s:%s\n", alias, name);
			*a = '.';
		}
}
